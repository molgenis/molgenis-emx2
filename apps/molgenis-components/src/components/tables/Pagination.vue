<template>
  <nav aria-label="Pagination">
    <ul class="pagination justify-content-center mb-0">
      <li class="page-item" :class="{ disabled: isFirstPage }">
        <a
          class="page-link"
          href="#"
          @click.prevent="emitValue(1, isFirstPage)"
        >
          <span aria-hidden="true">&laquo;</span>
          <span class="sr-only">First</span></a
        >
      </li>
      <li class="page-item" :class="{ disabled: isFirstPage }">
        <a
          class="page-link"
          href="#"
          @click.prevent="emitValue(Math.max(modelValue - 1, 1), isFirstPage)"
        >
          <span aria-hidden="true">&lsaquo;</span>
          <span class="sr-only">Previous</span>
        </a>
      </li>
      <li class="page-item disabled">
        <a class="page-link text-nowrap" href="#">
          {{ rowRange(modelValue, limit, count) }}
        </a>
      </li>
      <li class="page-item" :class="{ disabled: isLastPage }">
        <a
          class="page-link"
          href="#"
          @click.prevent="
            emitValue(Math.min(modelValue + 1, totalPages), isLastPage)
          "
        >
          <span aria-hidden="true">&rsaquo;</span>
          <span class="sr-only">Next</span></a
        >
      </li>
      <li class="page-item" :class="{ disabled: isLastPage }">
        <a
          class="page-link"
          href="#"
          @click.prevent="emitValue(totalPages, isLastPage)"
        >
          <span aria-hidden="true">&raquo;</span>
          <span class="sr-only">Last</span></a
        >
      </li>
    </ul>
  </nav>
</template>

<style scope>
.page-item {
  border: none;
}
</style>

<script lang="ts">
import { defineComponent } from "vue";

export default defineComponent({
  props: {
    /** current page */
    modelValue: { type: Number, default: 1 },
    /** total records, i.e. sql count */
    count: { type: Number, required: true },
    /** number of records per page, i.e. sql limit */
    limit: { type: Number, default: 10 },
  },
  methods: {
    emitValue(page: number, isDisabled: boolean) {
      if (!isDisabled) {
        this.$emit("update:modelValue", page);
      }
    },
    rowRange(value: number, limit: number, count: number) {
      if (count === 0) {
        return "-";
      } else {
        const from = (value - 1) * limit + 1;
        const to = Math.min(count, value * limit);
        return `${from} - ${to} of ${count}`;
      }
    },
  },
  computed: {
    totalPages() {
      return Math.ceil(this.count / this.limit);
    },
    isFirstPage() {
      return this.modelValue === 1;
    },
    isLastPage() {
      return this.modelValue === this.totalPages || this.count === 0;
    },
  },
  watch: {
    value() {
      if (this.modelValue < 1) {
        this.$emit("update:modelValue", 1);
      }
    },
    count() {
      //reset page to within range in case count changes
      if (this.modelValue > this.totalPages) {
        this.$emit("update:modelValue", 1);
      }
    },
  },
  created() {
    if (this.modelValue < 1) {
      this.$emit("update:modelValue", 1);
    }
  },
});
</script>

<docs>
<template>

  <div>
    <demo-item>
      <Pagination v-model="pageValue1" :count="29" />
      page = {{ pageValue1 }}
    </demo-item>

    <demo-item>
      <pagination v-model="pageValue2" :count="250" />
      <div>page number: {{ pageValue2 }}</div>
    </demo-item>
  </div>

</template>

<script>
  export default {
    data() {
      return {
        pageValue1: null,
        pageValue2: 3,
      }
    }
  }
</script>
</docs>
