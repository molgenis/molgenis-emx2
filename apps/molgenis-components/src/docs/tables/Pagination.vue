<template>
  <demo-item id="pagination" label="Pagination">
    <pagination v-model="page" :count="250" />
    <div>page number: {{ page }}</div>
  </demo-item>
</template>

<script>
export default {
  data() {
    return {
      page: 3,
    };
  },
};
</script>