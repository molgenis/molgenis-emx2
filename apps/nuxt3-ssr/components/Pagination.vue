<!-- eslint-disable vue/multi-word-component-names -->
<script setup>
import { computed } from "vue";

const props = defineProps({
  currentPage: {
    type: Number,
    required: true,
  },
  totalPages: {
    type: Number,
    required: true,
  },
  type: {
    type: String,
    default: "gray",
    enum: ["gray", "white"],
  },
  preventDefault: {
    type: Boolean,
    default: false,
  },
});
const emit = defineEmits(["update"]);

onMounted(() => {
  window.addEventListener("popstate", () => {
    // react to external navigation ( e.g. back button in browser)
    window.location.reload();
  });
});

const TEXT_STYLE_MAPPING = {
  gray: "text-pagination-label-gray",
  white: "text-pagination-label-white",
};

const BORDER_STYLE_MAPPING = {
  gray: "shadow-pagination-gray",
  white: "",
};

const textClasses = computed(() => {
  return TEXT_STYLE_MAPPING[props.type];
});

const borderClasses = computed(() => {
  return BORDER_STYLE_MAPPING[props.type];
});

function onPrevClick($event) {
  if (props.preventDefault) {
    $event.preventDefault();
  }
  if (props.currentPage > 1) {
    emit("update", props.currentPage - 1);
  }
}

function onNextClick($event) {
  if (props.preventDefault) {
    $event.preventDefault();
  }
  if (props.currentPage < props.totalPages) {
    emit("update", props.currentPage + 1);
  }
}

function changeCurrentPage(event) {
  const newPage = parseInt(event.target.value);
  const clampedPage = Math.min(Math.max(newPage, 1), props.totalPages);
  if (isNaN(clampedPage)) {
    emit("update", 1);
  } else {
    emit("update", clampedPage);
  }
}
</script>

<template>
  <nav
    class="pt-12.5 flex items-center justify-center font-display text-heading-xl -mx-2.5"
  >
    <a
      :href="currentPage > 1 ? '' : undefined"
      role="button"
      @click="onPrevClick"
      class="flex justify-center transition-colors border border-pagination rounded-pagination bg-pagination text-pagination h-15 w-15"
      :class="{
        'hover:bg-pagination-hover hover:text-pagination-hover':
          currentPage > 1,
      }"
    >
      <BaseIcon name="caret-left" :width="24" />
    </a>
    <div class="px-4 tracking-widest sm:px-5" :class="textClasses">Page</div>
    <input
      class="sm:px-12 px-7.5 w-32 text-center border rounded-pagination text-pagination-input h-15 flex items-center tracking-widest bg-white"
      :value="currentPage"
      @change="changeCurrentPage"
      :class="borderClasses"
    />
    <div
      class="px-4 tracking-widest sm:px-5 whitespace-nowrap"
      :class="textClasses"
    >
      OF {{ totalPages }}
    </div>
    <a
      :href="currentPage < totalPages ? '' : undefined"
      role="button"
      @click="onNextClick"
      class="flex justify-center transition-colors border border-pagination rounded-pagination bg-pagination text-pagination h-15 w-15"
      :class="{
        'hover:bg-pagination-hover hover:text-pagination-hover':
          currentPage < totalPages,
      }"
    >
      <BaseIcon name="caret-right" :width="24" />
    </a>
  </nav>
</template>
