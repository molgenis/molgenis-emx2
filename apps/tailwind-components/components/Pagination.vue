<script setup lang="ts">
const pageInputId = useId();

const props = withDefaults(
  defineProps<{
    currentPage: number;
    totalPages: number;
    preventDefault?: boolean;
  }>(),
  {
    preventDefault: false,
  }
);
const emit = defineEmits(["update"]);

onMounted(() => {
  if (window) {
    window.addEventListener("popstate", () => {
      // react to external navigation ( e.g. back button in browser)
      window.location.reload();
    });
  }
});

function onPrevClick($event: Event) {
  if (props.preventDefault) {
    $event.preventDefault();
  }
  if (props.currentPage > 1) {
    emit("update", props.currentPage - 1);
  }
}

function onNextClick($event: Event) {
  if (props.preventDefault) {
    $event.preventDefault();
  }
  if (props.currentPage < props.totalPages) {
    emit("update", props.currentPage + 1);
  }
}

function changeCurrentPage(event: Event) {
  const newPage = parseInt((event.target as HTMLInputElement)?.value);
  const clampedPage = Math.min(Math.max(newPage, 1), props.totalPages);
  if (isNaN(clampedPage)) {
    emit("update", 1);
  } else {
    emit("update", clampedPage);
  }
}
</script>

<template>
  <nav
    role="navigation"
    class="pt-12.5 font-display text-heading-xl -mx-2.5"
    :aria-labelledby="`${pageInputId}Label`"
  >
    <span :id="`${pageInputId}Label`" class="sr-only"
      >pagingation navigation</span
    >
    <ul class="flex items-center justify-center list-none">
      <li>
        <a
          @click.prevent="onPrevClick"
          class="flex justify-center transition-colors border border-pagination rounded-pagination bg-pagination text-pagination h-15 w-15"
          :class="{
            'hover:bg-pagination-hover hover:text-pagination-hover':
              currentPage > 1,
          }"
        >
          <span class="sr-only">Go to page {{ currentPage - 1 }}</span>
          <BaseIcon name="caret-left" :width="24" />
        </a>
      </li>
      <li class="flex justify-center items-center">
        <div class="px-4 tracking-widest sm:px-5">
          <label :for="pageInputId" class="sr-only">go to specific page</label>
          <span class="text-pagination">Page</span>
        </div>
        <input
          :id="pageInputId"
          class="sm:px-12 px-7.5 w-32 text-center border rounded-pagination text-pagination-input h-15 flex items-center tracking-widest bg-white"
          :value="currentPage"
          @change="changeCurrentPage"
        />
        <div class="px-4 tracking-widest sm:px-5 whitespace-nowrap">
          <span class="text-pagination">OF {{ totalPages }}</span>
        </div>
      </li>
      <li>
        <a
          @click.prevent="onNextClick"
          class="flex justify-center transition-colors border border-pagination rounded-pagination bg-pagination text-pagination h-15 w-15"
          :class="{
            'hover:bg-pagination-hover hover:text-pagination-hover':
              currentPage < totalPages,
          }"
        >
          <span class="sr-only">Go to page {{ currentPage + 1 }}</span>
          <BaseIcon name="caret-right" :width="24" />
        </a>
      </li>
    </ul>
  </nav>
</template>
