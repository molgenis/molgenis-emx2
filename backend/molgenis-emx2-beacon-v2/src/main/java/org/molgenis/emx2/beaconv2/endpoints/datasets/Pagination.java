package org.molgenis.emx2.beaconv2.endpoints.datasets;

import com.fasterxml.jackson.annotation.JsonAutoDetect;
import com.fasterxml.jackson.annotation.JsonInclude;

@JsonAutoDetect(fieldVisibility = JsonAutoDetect.Visibility.ANY)
public class Pagination {

  @JsonInclude(JsonInclude.Include.ALWAYS)
  private int skip = 0;

  @JsonInclude(JsonInclude.Include.ALWAYS)
  private int limit = 10;
}
