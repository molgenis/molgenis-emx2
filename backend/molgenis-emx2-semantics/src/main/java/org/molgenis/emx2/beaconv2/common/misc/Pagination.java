package org.molgenis.emx2.beaconv2.common.misc;

public class Pagination {
  private int skip = 0;
  private int limit = 10;
}
