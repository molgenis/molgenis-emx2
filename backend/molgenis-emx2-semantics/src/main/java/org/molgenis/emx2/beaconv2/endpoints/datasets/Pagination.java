package org.molgenis.emx2.beaconv2.endpoints.datasets;

import com.fasterxml.jackson.annotation.JsonAutoDetect;

@JsonAutoDetect(fieldVisibility = JsonAutoDetect.Visibility.ANY)
public class Pagination {

  // TODO default? get from request?
  private int skip = 0;
  private int limit = 10;
}
