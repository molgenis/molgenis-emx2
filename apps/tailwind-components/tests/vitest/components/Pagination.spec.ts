import { mount } from "@vue/test-utils";
import { expect, it } from "vitest";
import Pagination from "../../../components/Pagination.vue";

const wrapper = mount(Pagination, {
  props: {
    currentPage: 3,
    totalPages: 34,
    preventDefault: true,
    inverted: false,
    jumpToEdge: false,
  },
});

it("should render the pagination component (without jump to edge)", async () => {
  expect(wrapper.html()).matchSnapshot();
  expect(wrapper.html()).not.toContain("doublearrowleft");
  expect(wrapper.html()).not.toContain("doublearrowright");
});

it("should render the pagination component with jump to edge", async () => {
  const withEdge = mount(Pagination, {
    props: {
      currentPage: 3,
      totalPages: 34,
      preventDefault: true,
      inverted: false,
      jumpToEdge: true,
    },
  });
  expect(withEdge.html()).matchSnapshot();
  expect(withEdge.html()).not.toStrictEqual(wrapper.html());
  expect(withEdge.html()).toContain("doublearrowleft");
  expect(withEdge.html()).toContain("doublearrowright");
});

it("should emit update event on button click", async () => {
  const first = wrapper.find("a").find("span");
  await first.trigger("click");
  expect(wrapper.emitted()).toHaveProperty("update");
  expect(wrapper.emitted("update")?.[0]).toEqual([2]);
  const last = wrapper.findAll("a").at(1)?.find("span");
  await last?.trigger("click");
  expect(wrapper.emitted("update")?.[1]).toEqual([4]);
});

it("should not emit update event when on first page and clicking first", async () => {
  const firstPage = mount(Pagination, {
    props: {
      currentPage: 1,
      totalPages: 34,
      preventDefault: true,
      inverted: false,
      jumpToEdge: false,
    },
  });
  const first = firstPage.find("a").find("span");
  await first.trigger("click");
  expect(firstPage.emitted("update")).toBeUndefined();
});

it("should not emit update event when on last page and clicking last", async () => {
  const lastPage = mount(Pagination, {
    props: {
      currentPage: 34,
      totalPages: 34,
      preventDefault: true,
      inverted: false,
      jumpToEdge: false,
    },
  });
  const last = lastPage.findAll("a").at(1)?.find("span");
  await last?.trigger("click");
  expect(lastPage.emitted("update")).toBeUndefined();
});

it("should not emit update when jumpToEdge is true and clicking first on first page", async () => {
  const firstPage = mount(Pagination, {
    props: {
      currentPage: 1,
      totalPages: 34,
      preventDefault: true,
      inverted: false,
      jumpToEdge: true,
    },
  });
  const first = firstPage.findAll("a").at(0)?.find("span");
  await first?.trigger("click");
  expect(firstPage.emitted("update")).toBeUndefined();
});

it("should not emit update when jumpToEdge is true and clicking last on last page", async () => {
  const lastPage = mount(Pagination, {
    props: {
      currentPage: 34,
      totalPages: 34,
      preventDefault: true,
      inverted: false,
      jumpToEdge: true,
    },
  });
  const last = lastPage.findAll("a").at(3)?.find("span");
  await last?.trigger("click");
  expect(lastPage.emitted("update")).toBeUndefined();
});
