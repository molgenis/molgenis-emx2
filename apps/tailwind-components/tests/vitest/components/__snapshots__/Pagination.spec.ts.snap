// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should render the pagination component  1`] = `
"<nav role="navigation" class="pt-12.5 font-display text-heading-xl -mx-2.5" aria-labelledby="v-0Label"><span id="v-0Label" class="sr-only"> pagination navigation </span>
  <ul class="flex items-center justify-center list-none gap-2.5">
    <!--v-if-->
    <li><a href="#" class="flex justify-center border border-pagination rounded-theme bg-pagination text-pagination-button h-15 w-15 cursor-default cursor-pointer hover:bg-pagination-hover hover:text-pagination-hover hover:border-pagination-hover focus:bg-pagination-hover focus:text-pagination-hover"><span class="sr-only">Go to page 2</span>
        <caretleft width="24"></caretleft>
      </a></li>
    <li class="flex justify-center items-center">
      <div class="px-4 tracking-widest sm:px-5"><label for="v-0" class="sr-only">go to specific page</label><span class="text-pagination"> Page </span></div><input id="v-0" class="sm:px-12 px-7.5 w-32 text-center border border-input rounded-theme bg-input text-pagination-input hover:text-pagination-hover hover:border-pagination-hover hover:bg-pagination-hover h-15 flex items-center tracking-widest" value="3">
      <div class="px-4 tracking-widest sm:px-5 whitespace-nowrap"><span class="text-pagination"> OF 34</span></div>
    </li>
    <li><a href="#" class="flex justify-center border border-pagination rounded-theme bg-pagination text-pagination-button h-15 w-15 cursor-default cursor-pointer hover:bg-pagination-hover hover:text-pagination-hover hover:border-pagination-hover focus:bg-pagination-hover focus:text-pagination-hover"><span class="sr-only">Go to page 4</span>
        <caretright width="24"></caretright>
      </a></li>
    <!--v-if-->
  </ul>
</nav>"
`;

exports[`should render the pagination component (without jump to edge) 1`] = `
"<nav role="navigation" class="pt-12.5 font-display text-heading-xl -mx-2.5" aria-labelledby="v-0Label"><span id="v-0Label" class="sr-only"> pagination navigation </span>
  <ul class="flex items-center justify-center list-none gap-2.5">
    <!--v-if-->
    <li><a href="#" class="flex justify-center border border-pagination rounded-theme bg-pagination text-pagination-button h-15 w-15 cursor-default cursor-pointer hover:bg-pagination-hover hover:text-pagination-hover hover:border-pagination-hover focus:bg-pagination-hover focus:text-pagination-hover"><span class="sr-only">Go to page 2</span>
        <caretleft width="24"></caretleft>
      </a></li>
    <li class="flex justify-center items-center">
      <div class="px-4 tracking-widest sm:px-5"><label for="v-0" class="sr-only">go to specific page</label><span class="text-pagination"> Page </span></div><input id="v-0" class="sm:px-12 px-7.5 w-32 text-center border border-input rounded-theme bg-input text-pagination-input hover:text-pagination-hover hover:border-pagination-hover hover:bg-pagination-hover h-15 flex items-center tracking-widest" value="3">
      <div class="px-4 tracking-widest sm:px-5 whitespace-nowrap"><span class="text-pagination"> OF 34</span></div>
    </li>
    <li><a href="#" class="flex justify-center border border-pagination rounded-theme bg-pagination text-pagination-button h-15 w-15 cursor-default cursor-pointer hover:bg-pagination-hover hover:text-pagination-hover hover:border-pagination-hover focus:bg-pagination-hover focus:text-pagination-hover"><span class="sr-only">Go to page 4</span>
        <caretright width="24"></caretright>
      </a></li>
    <!--v-if-->
  </ul>
</nav>"
`;

exports[`should render the pagination component 1`] = `
"<nav role="navigation" class="pt-12.5 font-display text-heading-xl -mx-2.5" aria-labelledby="v-0Label"><span id="v-0Label" class="sr-only"> pagination navigation </span>
  <ul class="flex items-center justify-center list-none gap-2.5">
    <!--v-if-->
    <li><a href="#" class="flex justify-center border border-pagination rounded-theme bg-pagination text-pagination-button h-15 w-15 cursor-default cursor-pointer hover:bg-pagination-hover hover:text-pagination-hover hover:border-pagination-hover focus:bg-pagination-hover focus:text-pagination-hover"><span class="sr-only">Go to page 2</span>
        <caretleft width="24"></caretleft>
      </a></li>
    <li class="flex justify-center items-center">
      <div class="px-4 tracking-widest sm:px-5"><label for="v-0" class="sr-only">go to specific page</label><span class="text-pagination"> Page </span></div><input id="v-0" class="sm:px-12 px-7.5 w-32 text-center border border-input rounded-theme bg-input text-pagination-input hover:text-pagination-hover hover:border-pagination-hover hover:bg-pagination-hover h-15 flex items-center tracking-widest" value="3">
      <div class="px-4 tracking-widest sm:px-5 whitespace-nowrap"><span class="text-pagination"> OF 34</span></div>
    </li>
    <li><a href="#" class="flex justify-center border border-pagination rounded-theme bg-pagination text-pagination-button h-15 w-15 cursor-default cursor-pointer hover:bg-pagination-hover hover:text-pagination-hover hover:border-pagination-hover focus:bg-pagination-hover focus:text-pagination-hover"><span class="sr-only">Go to page 4</span>
        <caretright width="24"></caretright>
      </a></li>
    <!--v-if-->
  </ul>
</nav>"
`;

exports[`should render the pagination component with jump to edge 1`] = `
"<nav role="navigation" class="pt-12.5 font-display text-heading-xl -mx-2.5" aria-labelledby="v-0Label"><span id="v-0Label" class="sr-only"> pagination navigation </span>
  <ul class="flex items-center justify-center list-none gap-2.5">
    <li><a href="#" class="flex justify-center border border-pagination rounded-theme bg-pagination text-pagination-button h-15 w-15 cursor-default cursor-pointer hover:bg-pagination-hover hover:text-pagination-hover hover:border-pagination-hover focus:bg-pagination-hover focus:text-pagination-hover"><span class="sr-only">Go to page first</span>
        <doublearrowleft width="24"></doublearrowleft>
      </a></li>
    <li><a href="#" class="flex justify-center border border-pagination rounded-theme bg-pagination text-pagination-button h-15 w-15 cursor-default cursor-pointer hover:bg-pagination-hover hover:text-pagination-hover hover:border-pagination-hover focus:bg-pagination-hover focus:text-pagination-hover"><span class="sr-only">Go to page 2</span>
        <caretleft width="24"></caretleft>
      </a></li>
    <li class="flex justify-center items-center">
      <div class="px-4 tracking-widest sm:px-5"><label for="v-0" class="sr-only">go to specific page</label><span class="text-pagination"> Page </span></div><input id="v-0" class="sm:px-12 px-7.5 w-32 text-center border border-input rounded-theme bg-input text-pagination-input hover:text-pagination-hover hover:border-pagination-hover hover:bg-pagination-hover h-15 flex items-center tracking-widest" value="3">
      <div class="px-4 tracking-widest sm:px-5 whitespace-nowrap"><span class="text-pagination"> OF 34</span></div>
    </li>
    <li><a href="#" class="flex justify-center border border-pagination rounded-theme bg-pagination text-pagination-button h-15 w-15 cursor-default cursor-pointer hover:bg-pagination-hover hover:text-pagination-hover hover:border-pagination-hover focus:bg-pagination-hover focus:text-pagination-hover"><span class="sr-only">Go to page 4</span>
        <caretright width="24"></caretright>
      </a></li>
    <li><a href="#" class="flex justify-center border border-pagination rounded-theme bg-pagination text-pagination-button h-15 w-15 cursor-default cursor-pointer hover:bg-pagination-hover hover:text-pagination-hover hover:border-pagination-hover focus:bg-pagination-hover focus:text-pagination-hover"><span class="sr-only">Go to last page</span>
        <doublearrowright width="24"></doublearrowright>
      </a></li>
  </ul>
</nav>"
`;
